### PR TITLE
Add chacha20/poly1305 and chacha20poly1305_AEAD from openssh

### PR DESCRIPTION
**Pull request by jonasschnelli**
This adds **ChaCha20** and **Poly1305** and the openSSH form of the AEAD.

The implementations of **ChaCha20** are from DJB himself and directly taken (unchanged as much as possible) from the openSSH source code: https://github.com/openssh/openssh-portable/blob/master/chacha.c

**Poly1305** is written by Andrew Moon is also directly taken (unchanged as much as possible) from the openSSH source code: https://github.com/openssh/openssh-portable/blob/master/poly1305.c

The code for the **ChaCha20/Poly1305@openssh** AEAD is taken from https://github.com/openssh/openssh-portable/blob/master/cipher-chachapoly.c with minimal changes.

More details on the AEAD construction:
https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.chacha20poly1305
https://gist.github.com/jonasschnelli/c530ea8421b8d0e80c51486325587c52

Test vectors are taken from the RFC draft:
https://tools.ietf.org/html/draft-agl-tls-chacha20poly1305-04#section-7

Taking existing implementations should reduce risks of screwing up at the implementation level.

**OUT OF SCOPE**:
* AVX,SSE acceleration (possible, see libsodium a.s.o.)
* To Bitcoin Core adjusted C++ implementations
* Combining with the existing ChaCha20 RNG
 (can all be done later, after verification of the correctness)

This is a subset of #14032 and a pre-requirement for BIP151.

### CODE REVIEW

**Comment by DesWurstes** [comment]
Why not [reuse the existing ChaCha20](https://github.com/bitcoin/bitcoin/compare/master...DesWurstes:master)?

**Comment by jonasschnelli** [comment]
@DesWurstes: see PR description (out of scope).

**Comment by jonasschnelli** [comment]
Added benchmark for the AEAD (and for direct comparison also added the dbl-SHA (HASH) benchmark).
Benchmarking 1MB data as well as 256bytes.

My system reports (!we compare SSE4 SHA256 vs non-optimized chacha20):
```
# Benchmark, evals, iterations, total, min, max, median
CHACHA20POLY1305AEAD_BIG, 5, 340, 3.68279, 0.00215035, 0.00219169, 0.00216025
CHACHA20POLY1305AEAD_SMALL, 5, 250000, 1.08673, 8.51516e-07, 8.93585e-07, 8.61119e-07
HASH256_BIG, 5, 340, 3.81384, 0.00222589, 0.00226436, 0.00224086
HASH256_SMALL, 5, 250000, 1.1305, 8.96669e-07, 9.15482e-07, 9.03866e-07
```

File: src/crypto/chachapoly_aead.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+void _memory_cleanse(void *ptr, size_t len)
+{
+    memset(ptr, 0, len);
+
+    /* As best as we can tell, this is sufficient to break any optimisations that
+       might try to eliminate "superfluous" memsets. If there's an easy way to
+       detect memset_s, it would be better to use that. */
+#if defined(_MSC_VER)
+    SecureZeroMemory(ptr, len);
```
Comment (ken2812221) [inline src/crypto/chachapoly_aead.cpp]: Seems this should include windows.h

File: src/crypto/chachapoly_aead.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  struct chacha_ctx main_ctx, header_ctx;
+};
+
+int chacha20poly1305_init(struct chachapolyaead_ctx *cpctx, const uint8_t *key,
+                          int keylen);
+int chacha20poly1305_crypt(struct chachapolyaead_ctx *ctx, uint32_t seqnr,
+                           uint8_t *dest, const uint8_t *src, uint32_t len,
+                           uint32_t aadlen, int is_encrypt);
+
+// extracts the LE 24bit (3byte) length from the AAD and puts it into a uint32_t (host endiannes)
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.h]: Typo found by `codespell`: endiannes

Comment (jonasschnelli) [reply src/crypto/chachapoly_aead.h]: Fixed.

File: src/test/crypto_tests.cpp [hunk 3]
```diff
@@ -562,4 +564,198 @@ BOOST_AUTO_TEST_CASE(sha256d64)
...
+        uint8_t resulting_tag[16];
+    };
+
+    /*
+       Testvectors have been taken from the draft RFC
+       https://tools.ietf.org/html/draft-agl-tls-chacha20poly1305-04#section-7
+    */
+
+    static const struct chacha20_testvector chacha20_testvectors[] = {
+        {{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
```
Comment (Sjors) [inline src/test/crypto_tests.cpp]: This would be much easier to verify against the RFC document if you parse a hex string.

File: src/crypto/poly1305.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+#define U32TO8_LE(p, v)                                                        \
+  do {                                                                         \
+    (p)[0] = (uint8_t)((v));                                                   \
+    (p)[1] = (uint8_t)((v) >> 8);                                              \
+    (p)[2] = (uint8_t)((v) >> 16);                                             \
+    (p)[3] = (uint8_t)((v) >> 24);                                             \
+  } while (0)
+
+void poly1305_auth(unsigned char out[POLY1305_TAGLEN], const unsigned char *m,
```
Comment (Sjors) [inline src/crypto/poly1305.cpp]: Why did you remove the [poly1305_init](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna-32.h), [poly1305_update](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna.c#L27), [poly1305_finish](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna-32.h#L134) separation?

File: src/crypto/chacha.cpp [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  x->input[9] = U8TO32_LITTLE(k + 4);
+  x->input[10] = U8TO32_LITTLE(k + 8);
+  x->input[11] = U8TO32_LITTLE(k + 12);
+  x->input[0] = U8TO32_LITTLE(constants + 0);
+  x->input[1] = U8TO32_LITTLE(constants + 4);
+  x->input[2] = U8TO32_LITTLE(constants + 8);
+  x->input[3] = U8TO32_LITTLE(constants + 12);
+}
+
+void chacha_ivsetup(chacha_ctx *x, const u8 *iv, const u8 *counter) {
```
Comment (Sjors) [inline src/crypto/chacha.cpp]: I can confirm this function in 09ec49ba31867187cecd631593d37bcc5f3ca72f is identical to the original at de624c626ea081929df000b09932dbc804eda51d.

**Review by Sjors — COMMENTED** [review]
Re out of scope "Combining with the existing ChaCha20 RNG": perhaps you can add some tests to show they're identical?

File: src/crypto/poly1305.cpp (hunk 4 above)
Comment (jonasschnelli) [reply src/crypto/poly1305.cpp]: Which separation? I took Andrew Moons implementation found at https://github.com/openssh/openssh-portable/blob/90e51d672711c19a36573be1785caf35019ae7a8/poly1305.c

**Comment by jonasschnelli** [comment]
@Sjors The existing ChaCha20 implementation doesn't do the XORing and can therefore only be used for the RNG.

File: src/crypto/poly1305.cpp (hunk 4 above)
Comment (Sjors) [reply src/crypto/poly1305.cpp]: Ah, I see, in that case the source code URL at the top of the file is confusing.

**Comment by sipa** [comment]
Instead of those U8TOU32_LE macros you can use our own function (ReadLE32,
WriteLE32).

>

**Comment by jonasschnelli** [comment]
@sipa: agree. Though as written in the PR description. I'd like to keep the code (extracted from openSSH) as identical as possible. I think (or hope) we can do refactoring after this has been merged.

File: src/crypto/chachapoly_aead.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+int chacha20poly1305_get_length24(struct chachapolyaead_ctx *ctx,
+                                uint32_t *len24_out, uint32_t seqnr,
+                                const uint8_t *ciphertext) {
+  uint8_t buf[3];
+
+  uint64_t seqnr64 = htole64(seqnr); //use LE for the nonce
+  chacha_ivsetup(&ctx->header_ctx, (uint8_t *)&seqnr64, NULL);
+  chacha_encrypt_bytes(&ctx->header_ctx, ciphertext, buf, 3);
+  *len24_out = 0;
+  *len24_out = buf[0] | (buf[1] << 8) | (buf[2] << 16);
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.cpp]: This seems like a mistake? `*len24_out` is assigned twice? `*len24_out = 0;` should be removed? :-)

File: src/crypto/poly1305.cpp [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+  b = (g4 >> 31) - 1;
+  nb = ~b;
+  h0 = (h0 & nb) | (g0 & b);
+  h1 = (h1 & nb) | (g1 & b);
+  h2 = (h2 & nb) | (g2 & b);
+  h3 = (h3 & nb) | (g3 & b);
+  h4 = (h4 & nb) | (g4 & b);
+
+  f0 = ((h0) | (h1 << 26)) + (uint64_t)U8TO32_LE(&key[16]);
```
Comment (practicalswift) [inline src/crypto/poly1305.cpp]: Just to double check: This bit shifting takes place on a 32-bit value which is then expanded to a 64-bit type. That is the intention? Applies to `h1 << 26`, `h2 << 20`, `h3 << 14` and `h4 << 8`.

File: src/test/crypto_tests.cpp [hunk 8]
```diff
@@ -562,4 +564,198 @@ BOOST_AUTO_TEST_CASE(sha256d64)
...
+    chacha20poly1305_init(&aead_ctx, aead_keys, 64);
+    assert((uint32_t)plaintext_buf[0] == 255);
+    chacha20poly1305_crypt(&aead_ctx, seqnr, ciphertext_buf, plaintext_buf, 252,
+        3, 1);
+    uint32_t out_len = 0;
+    chacha20poly1305_get_length24(&aead_ctx, &out_len, seqnr, ciphertext_buf);
+    assert(out_len == 255);
+    chacha20poly1305_crypt(&aead_ctx, seqnr, plaintext_buf_new, ciphertext_buf,
+        252, 3, 0);
+    assert(memcmp(plaintext_buf, plaintext_buf_new, 252) == 0);
```
Comment (practicalswift) [inline src/test/crypto_tests.cpp]: Shouldn't this be `assert(memcmp(plaintext_buf, plaintext_buf_new, sizeof(plaintext_buf)) == 0);`?

File: src/bench/chacha20poly1305.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    static const uint8_t aead_keys[64] = {
+        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a,
+        0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15,
+        0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+    chacha20poly1305_init(&aead_ctx, aead_keys, 64);
+    std::vector<uint8_t> in(bufsize,0);
```
Comment (practicalswift) [inline src/bench/chacha20poly1305.cpp]:
```
2018-09-22 21:20:48 cpplint(pr=14050): src/bench/chacha20poly1305.cpp:29:  Missing space after ,  [whitespace/comma] [3]
```

File: src/bench/chacha20poly1305.cpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    std::vector<uint8_t> out(bufsize,0);
```
Comment (practicalswift) [inline src/bench/chacha20poly1305.cpp]:
```
2018-09-22 21:20:48 cpplint(pr=14050): src/bench/chacha20poly1305.cpp:30:  Missing space after ,  [whitespace/comma] [3]
```

File: src/bench/chacha20poly1305.cpp [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+static void CHACHA20POLY1305AEAD_SMALL(benchmark::State& state)
+{
+    CHACHA20POLY1305AEAD_(state, BUFFER_SIZE_B);
+}
+
+static void HASH256_(benchmark::State& state, uint64_t bufsize)
+{
+    uint8_t hash[CHash256::OUTPUT_SIZE];
+    std::vector<uint8_t> in(bufsize,0);
```
Comment (practicalswift) [inline src/bench/chacha20poly1305.cpp]:
```
2018-09-22 21:20:48 cpplint(pr=14050): src/bench/chacha20poly1305.cpp:51:  Missing space after ,  [whitespace/comma] [3]
```

File: src/crypto/chachapoly_aead.cpp [hunk 12]
```diff
@@ -0,0 +1,90 @@
+#include <crypto/chachapoly_aead.h>
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.cpp]:
```
2018-09-22 21:20:48 cpplint(pr=14050): src/crypto/chachapoly_aead.cpp:0:  No copyright message found.
```

File: src/crypto/chachapoly_aead.cpp [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+  memory_cleanse(poly_key, sizeof(poly_key));
+  return r;
+}
+
+int chacha20poly1305_get_length24(struct chachapolyaead_ctx *ctx,
+                                uint32_t *len24_out, uint32_t seqnr,
+                                const uint8_t *ciphertext) {
+  uint8_t buf[3];
+
+  uint64_t seqnr64 = htole64(seqnr); //use LE for the nonce
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.cpp]:
```
2018-09-22 21:20:48 cpplint(pr=14050): src/crypto/chachapoly_aead.cpp:81:  Should have a space between // and comment  [whitespace/comments] [4]
```

File: src/crypto/chachapoly_aead.h [hunk 14]
```diff
@@ -0,0 +1,22 @@
+#ifndef BITCOIN_CRYPTO_CHACHAPOLY_AEAD_H
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.h]:
```
2018-09-22 21:20:48 cpplint(pr=14050): src/crypto/chachapoly_aead.h:0:  No copyright message found.
```

File: src/bench/chacha20poly1305.cpp [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+    chacha20poly1305_init(&aead_ctx, aead_keys, 64);
+    std::vector<uint8_t> in(bufsize,0);
+    std::vector<uint8_t> out(bufsize,0);
+
+    while (state.KeepRunning()) {
+        chacha20poly1305_crypt(&aead_ctx, seqnr, in.data(), out.data(), in.size()-16,
```
Comment (practicalswift) [inline src/bench/chacha20poly1305.cpp]:
```
2018-09-25 21:29:15 clang(pr=14050): bench/chacha20poly1305.cpp:33:82: warning: implicit conversion loses integer precision: 'unsigned long' to 'uint32_t' (aka 'unsigned int') [-Wshorten-64-to-32]
```

File: src/test/crypto_tests.cpp [hunk 16]
```diff
@@ -562,4 +564,198 @@ BOOST_AUTO_TEST_CASE(sha256d64)
...
+    struct chacha_ctx ctx;
+    unsigned int i = 0;
+    uint8_t keystream[512];
+    uint8_t poly1305_tag[16];
+
+    /* test chacha20 */
+    for (i = 0;
+         i < (sizeof(chacha20_testvectors) / sizeof(chacha20_testvectors[0]));
+         i++) {
+        chacha_ivsetup(&ctx, chacha20_testvectors[i].nonce, NULL);
```
Comment (practicalswift) [inline src/test/crypto_tests.cpp]:
```
2018-09-25 21:29:15 clang(pr=14050): test/crypto_tests.cpp:690:61: warning: zero as null pointer constant [-Wzero-as-null-pointer-constant]
```

Use `nullptr` :-)

File: src/crypto/poly1305.cpp [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+  g1 = h1 + b;
+  b = g1 >> 26;
+  g1 &= 0x3ffffff;
+  g2 = h2 + b;
+  b = g2 >> 26;
+  g2 &= 0x3ffffff;
+  g3 = h3 + b;
+  b = g3 >> 26;
+  g3 &= 0x3ffffff;
+  g4 = h4 + b - (1 << 26);
```
Comment (practicalswift) [inline src/crypto/poly1305.cpp]: An unsigned integer wraparound takes place here. Perhaps worth adding a comment about that to make it clear it is intentional?

File: src/crypto/chachapoly_aead.cpp [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+  chacha_keysetup(&ctx->header_ctx, key + 32, 256);
+  return 0;
+}
+
+int chacha20poly1305_crypt(struct chachapolyaead_ctx *ctx, uint32_t seqnr,
+                           uint8_t *dest, const uint8_t *src, uint32_t len,
+                           uint32_t aadlen, int is_encrypt) {
+  const uint8_t one[8] = {1, 0, 0, 0, 0, 0, 0, 0}; /* NB little-endian */
+  uint8_t expected_tag[POLY1305_TAGLEN], poly_key[POLY1305_KEYLEN];
+  int r = -1;
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.cpp]: This is a dead store :-)

File: src/crypto/chachapoly_aead.cpp [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+  uint64_t seqnr64 = seqnr;
+  uint64_t chacha_iv = htole64(seqnr64);
+  memset(poly_key, 0, sizeof(poly_key));
+  chacha_ivsetup(&ctx->main_ctx, (uint8_t *)&chacha_iv, NULL);
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.cpp]: Use `nullptr` :-)

File: src/crypto/chachapoly_aead.cpp [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+    poly1305_auth(expected_tag, src, aadlen + len, poly_key);
+    if (timingsafe_bcmp(expected_tag, tag, POLY1305_TAGLEN) != 0) {
+      r = -1;
+      goto out;
+    }
+  }
+
+  if (aadlen) {
+    chacha_ivsetup(&ctx->header_ctx, (uint8_t *)&chacha_iv, NULL);
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.cpp]: Use `nullptr` :-)

File: src/crypto/chachapoly_aead.cpp [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+  return r;
+}
+
+int chacha20poly1305_get_length24(struct chachapolyaead_ctx *ctx,
+                                uint32_t *len24_out, uint32_t seqnr,
+                                const uint8_t *ciphertext) {
+  uint8_t buf[3];
+
+  uint64_t seqnr64 = htole64(seqnr); //use LE for the nonce
+  chacha_ivsetup(&ctx->header_ctx, (uint8_t *)&seqnr64, NULL);
```
Comment (practicalswift) [inline src/crypto/chachapoly_aead.cpp]: Use `nullptr` :-)

File: src/crypto/poly1305.cpp (hunk 17 above)
Comment (jonasschnelli) [reply src/crypto/poly1305.cpp]: Its upstream code from openssh,.. the plan is to refactor this to c++ after initial merge to master.

File: src/crypto/chachapoly_aead.cpp (hunk 19 above)
Comment (jonasschnelli) [reply src/crypto/chachapoly_aead.cpp]: I'd like to keep this as C code for now, the .cpp filename extension is for bypassing the CXXFLAGS/CFLAGS issue.

File: src/crypto/chachapoly_aead.cpp (hunk 6 above)
Comment (jonasschnelli) [reply src/crypto/chachapoly_aead.cpp]: I'm not sure if applying 3 time a uint_8t (24bits) the way its done here will guarantee to set the other 8 bits to 0. This is why I initialise with 0, then set bit 0 to 23. Could be that I'm wrong though.

File: src/crypto/poly1305.cpp (hunk 7 above)
Comment (jonasschnelli) [reply src/crypto/poly1305.cpp]: Looks like.
This has been taken from openSSH (https://github.com/openssh/openssh-portable/blob/90e51d672711c19a36573be1785caf35019ae7a8/poly1305.c).

**Comment by jonasschnelli** [comment]
Fixed points reported by @practicalswift

File: src/crypto/chachapoly_aead.cpp (hunk 6 above)
Comment (practicalswift) [reply src/crypto/chachapoly_aead.cpp]: No need to initialise before assignment if the RHS evaluates to an integer, right?

`buf[0] | (buf[1] << 8) | (buf[2] << 16)` evaluates to an integer.

But don't trust me – I'm just a fellow human.

Let's verify using our favourite C++ interpreter `cling`:

```
$ cling

****************** CLING ******************
* Type C++ code and press enter to run it *
*             Type .q to exit             *
*******************************************
[cling]$ typeid(1).name()
(const char *) "i"
[cling]$ #include <cstdint>
[cling]$ uint8_t buf[3] = {1, 2, 3};
[cling]$ typeid(buf[0] | (buf[1] << 8) | (buf[2] << 16)).name()
(const char *) "i"
[cling]$ buf[0] | (buf[1] << 8) | (buf[2] << 16)
(int) 197121
```

:-)

**Comment by jonasschnelli** [comment]
Partially superseded by #15512 (the ChaCha20 part)

**Comment by jonasschnelli** [comment]
Closing...
It's now replaced with #15512,  #15519 and  #15649.
